Refactor the drawing of TCO's; Get rid of hardcoded colors in TCOs; Even out the color scheme

Umcaruje opened the pull request:
Per https://github.com/HDDigitizerMusic/lmms-alt-theme/issues/8

Fixes #2148

This PR does several things:

1. Fixes incorrect rendering of TCO's: they was a one pixel gap between the grid and the TCOs. Also fix the incorect rendering of the bar-lines on Instrument pattern TCO's:
Before & after
![screenshot from 2016-02-16 22 36 26](https://cloud.githubusercontent.com/assets/6282045/13092401/031899b6-d500-11e5-9cf4-535731f54c30.png)     ![screenshot from 2016-02-16 22 36 48](https://cloud.githubusercontent.com/assets/6282045/13092402/031bab4c-d500-11e5-85b2-c80b8aae5cca.png)

2. Adds a Boolean `qproperty-gradient` that allows themers to get rid of the hardcoded gradients on TCO's. 

3. Removes all of the hard coded colors, so now you can change the color of a muted and selected TCO, as well as the shadow of the text.

4. Evens out the muted and selected colors, so they all are the same.

5. Implements missing muted foreground color for automation patterns.

6. All pattern text stays the same color when the pattern is muted, for better readability.

Comparison:

Old
![screenshot from 2016-02-16 22 30 25](https://cloud.githubusercontent.com/assets/6282045/13092345/b5363c62-d4ff-11e5-8bdc-5f864a407c2f.png)

New
![screenshot from 2016-02-16 23 27 32](https://cloud.githubusercontent.com/assets/6282045/13093364/26ede986-d505-11e5-977e-820a9ebb7212.png)

New with gradients off
![screenshot from 2016-02-16 23 28 28](https://cloud.githubusercontent.com/assets/6282045/13093369/2aaec4c8-d505-11e5-8182-e18100162ed9.png)

IvanMaldonado commented:
@Umcaruje I don't have words to express my gratitude on the work you have done so far.
So many ideas for new themes with the new posibilites I can't wait for 1.2 to come out! Thanks!

Umcaruje commented:
@zonkmachine @tresf @michaelgregorius can I get some testing/review on this one?

tresf commented on data/themes/default/style.css:
```diff
@@ -540,28 +540,47 @@ TrackContainerView QLabel
+	qproperty-mutedColor: rgb( 80, 80, 80 );
+	qproperty-selectedColor: rgb( 0, 125, 255 );
 	qproperty-fgColor: rgb( 187, 227, 236 );
+	qproperty-fgMutedColor: rgb( 128, 128, 128 );
```
Of no implication to you... I'm not a fan of the `fgFoo` tag, personally because normally in css `fg` is synonymous with `foreground` which is synonymous with `color` which is synonymous with `textColor`.  In this case, it's the gradient/overlay, right? Long term, I think we need to cleanup our naming conventions a bit, but you've at least kept it consistent. :+1:

Umcaruje replied:
no, this is the color of the drawn notes/automation/waveform when a pattern is muted.

Umcaruje replied:
I was thinking about this. Maybe we can avoid a lot of this by using the `background-color` property in CSS. So, `fgColor` becomes `color` and `color` becomes `background-color`.

Also we could avoid the muted and selected properties by adding custom css pseudo states like
`AutomationPatternView:muted` or `AutomationPatternView:selected`

tresf replied:
Is `:muted` a valid selector?  If so, that may certainly help.  I'm glad you're thinking about it thought because we do need to cleanup and standardize the `style.css` eventually, and you've quickly become the most active dev looking at that file. :+1:

tresf commented on include/Track.h:
```diff
@@ -192,7 +192,12 @@ class TrackContentObjectView : public selectableObject, public ModelView
+	Q_PROPERTY( QColor selectedColor READ selectedColor WRITE setSelectedColor )
 	Q_PROPERTY( QColor textColor READ textColor WRITE setTextColor )
+	Q_PROPERTY( QColor textShadowColor READ textShadowColor WRITE setTextShadowColor )
+	Q_PROPERTY( bool gradient READ gradient WRITE setGradient )
```
I think I remember discussing this so sorry if I've switched gears, but is a `bool` necessary here?  Can't we simply set the alpha channel to `0` on the necessary colors?

Umcaruje replied:
? The colors get manualy converted into a linear gradient. If we set its alpha to 0 you would get a glitched pattern: 
![screenshot from 2016-02-15 23 36 50](https://cloud.githubusercontent.com/assets/6282045/13184791/2a77beb6-d73e-11e5-84db-6ea364b689ca.png)
Since there isn't anything drawn underneath it.

tresf replied:
Yes, I remember you making that point previously, thanks for reiterating it. :+1:

tresf commented on src/gui/AutomationPatternView.cpp:
```diff
@@ -270,17 +269,22 @@ void AutomationPatternView::paintEvent( QPaintEvent * )
+	} else {
+		p.setBrush( c );
+	}
+	
```
Again, of no consequence to you, but this logic could really be cleaned up... Food for thought. :beers: 

```diff
- if( m_needsUpdate == false )
-	{
-		QPainter p( this );
-		p.drawPixmap( 0, 0, m_paintPixmap );
-		return;
-	}

+ QPainter painter(this);
+
+ if (!m_needsUpdate)
+ {
+    painter.drawPixmap(0, 0, m_paintPixmap);
+    return;
+ }

-	QPainter _p( this );
-	const QColor styleColor = _p.pen().brush().color();
-
	m_needsUpdate = false;

-	if( m_paintPixmap.isNull() == true || m_paintPixmap.size() != size() )
-	{
-		m_paintPixmap = QPixmap( size() );
-	}

+	// init if null/empty
+	m_paintPixmap = m_paintPixmap.isNull() == true || m_paintPixmap.size() != size() ? QPixmap(size()) : m_paintPixmap;

	QPainter p( &m_paintPixmap );

	QLinearGradient lingrad( 0, 0, 0, height() );
	QColor c;

-	if( !( m_pat->getTrack()->isMuted() || m_pat->isMuted() ) )
-		c = styleColor;
-	else
-		c = mutedColor();
-
-	if( isSelected() == true )
-	{
-		c = selectedColor();
-	}

+	// priority:  selected, muted, normal
+	bool muted = m_pat->getTrack()->isMuted() || m_pat->isMuted();
+	c = isSelected() ? selectedColor() : (muted ? mutedColor() : painter.pen().brush().color());

	lingrad.setColorAt( 1, c.darker( 300 ) );
	lingrad.setColorAt( 0, c );
	
-	if ( gradient() ) {
-		p.setBrush( lingrad );
-	} else {
-		p.setBrush( c );
-	}

+	p.setBrush(gradient() ? lingrad : c);
```

Umcaruje replied:
>p.setBrush(gradient() ? lingrad : c);

Already tried that, and it is not allowed for the operands to be of a different type (QColor vs QLinearGradient)

I'll look into fixing the other stuff

tresf replied:
> not allowed for the operands to be of a different type (QColor vs QLinearGradient)

Ah yes... :+1:

Umcaruje replied:
How about putting the whole if statement into one line?

`if( gradient() ) { p.setBrush( lingrad ); } else { p.setBrush( c ); }`

tresf commented on src/tracks/Pattern.cpp:
```diff
@@ -1196,15 +1198,11 @@ void PatternView::paintEvent( QPaintEvent * )
-		p.setPen( QColor( 0, 0, 0 ) );
+		p.setPen( textShadowColor() );
 		p.drawText( 4, p.fontMetrics().height()+1, m_pat->name() );
-		p.setPen( text_color );
```
Probably benign, but can you explain the before and after here in regards to the way muted text color was/is handled?

Umcaruje replied:
Well, on some patterns the text changed the color when it was muted, and on some it remained the same. The text got pretty unreadable on muted patterns, and I think that's undesired, so I made the text always the same color, without looking if the pattern is muted or not. I noted it in the PR description: 
> 6. All pattern text stays the same color when the pattern is muted, for better readability.

Umcaruje replied:
Before and After: 
![screenshot from 2016-02-19 19 36 42](https://cloud.githubusercontent.com/assets/6282045/13185221/3fd627aa-d740-11e5-9b87-d6edc6a7431d.png) ![screenshot from 2016-02-19 19 37 21](https://cloud.githubusercontent.com/assets/6282045/13185220/3fd57bc0-d740-11e5-8dad-89529cc37892.png)

tresf replied:
> Before and After: 

:+1:

tresf commented:
Although you didn't write it, this draw code is bloated in my opinion.  Although not fair to pick on you about it, this may be our best chance to clean it up and make it consistent.  I left an example of what I mean here: https://github.com/LMMS/lmms/pull/2574#discussion_r53465923 but just about every draw function could benefit from some cleanup.

tresf commented on src/gui/AutomationPatternView.cpp:
```diff
@@ -332,9 +343,11 @@ void AutomationPatternView::paintEvent( QPaintEvent * )
-
-			p.fillRect( QRectF( x1, 0.0f, x2-x1, it.value() ),
-								lin2grad );
+			if ( gradient() ) {
```
Although I quite prefer `if () {` to `if () <newline> {`, we should try to be consistent with the styling of the current file.

Umcaruje replied:
Ah, ok, will fix.

Umcaruje commented:
@tresf I simplified the code, and also refactored it, so now all patterns draw in the same way/order. I also fixed the problem where the text was getting drawn over the pattern borders. I also commented the code in some places.

Inspired by #2579, I will implement static text drawing on the patterns, since their text doesn't change so often, which should speed up the gui.

tresf commented on src/gui/AutomationPatternView.cpp:
```diff
@@ -21,12 +21,12 @@
  * Boston, MA 02110-1301 USA.
  *
  */
+#include "AutomationPatternView.h"
```
@Umcaruje moving this up like this... is this just a "THIS IS MY HEADER!!" convention?  This is new to me.

Umcaruje replied:
Its in https://github.com/LMMS/lmms/wiki/Coding-conventions

>The exception is that in a source file, __the first included file should always be its header__.

tresf commented:
> @tresf I simplified the code, and also refactored it, so now all patterns draw in the same way/order. I also fixed the problem where the text was getting drawn over the pattern borders. I also commented the code in some places.

I think this is in a very good state now.  Since we use similar methods to draw it simplifies the code as well as the maintenance of it.

> Inspired by #2579, I will implement static text drawing on the patterns, since their text doesn't change so often, which should speed up the gui.

Good catch.

Have we put any thought into consolidation?  If we can consolidate the paint-event code, we may be able to make subclass-specific paint functions that pertain exactly to that particular class and handle everything else in the inherited class.  Likely not worth it this time around, but I wanted to mention it.... if we can draw all `Track` objects the same and then tweak the drawing of each subclass for its own particular type, it would further simplify this.  I've never done this in C++, so what I'm saying it purely hypothetical. :beers:

Umcaruje commented:
>Have we put any thought into consolidation? If we can consolidate the paint-event code, we may be able to make subclass-specific paint functions that pertain exactly to that particular class and handle everything else in the inherited class. Likely not worth it this time around, but I wanted to mention it.... if we can draw all Track objects the same and then tweak the drawing of each subclass for its own particular type, it would further simplify this. I've never done this in C++, so what I'm saying it purely hypothetical. 

The thing is, in `PatternView` we don't just draw the pattern, we also draw the BB editor (which is bad imo, but I won't be fixing that). Also the drawing of the bar-lines also differs from pattern to pattern, and the only thing connecting them all is the drawing of the fill, borders and text. It might be a good idea to consolidate, but that is out of the scope of this PR.

tresf commented:
:+1:

tresf commented on src/tracks/SampleTrack.cpp:
```diff
@@ -406,22 +399,27 @@ void SampleTCOView::paintEvent( QPaintEvent * _pe )
 		p.setFont( pointSize<7>( p.font() ) );
 
-		p.setPen( QColor( 0, 0, 0 ) );
+		p.setPen( textShadowColor() );
```
:+1:  for fixing future code too.

Umcaruje commented:
Update: `QStaticText` implemented. I also turned on antialiasing on the pattern text, which should make it a lot clearer (cc @MusikBear). I also made the B&B patterns render to a pixmap like the other ones. 
Also aligned up the sample tracks so they render perfectly in the center.

Comparison between old and new text: 
![screenshot from 2016-02-26 23 47 17](https://cloud.githubusercontent.com/assets/6282045/13367482/4a383df2-dce3-11e5-80ad-616bbbf5b0b7.png)
![screenshot from 2016-02-26 23 43 40](https://cloud.githubusercontent.com/assets/6282045/13367481/4a3517c6-dce3-11e5-97a6-32e7170b5924.png)

What's left to do: refactor the bar-line drawing code and also make sample tracks render to a pixmap.

Umcaruje commented:
@LMMS/developers I need a little help here. I reworked the `BBTCOView` to paint to a `QPixmap`, like `PatternView` and `AutomationPatternView` do, but I ran into a problem: 
![screenshot from 2016-02-27 00 06 39](https://cloud.githubusercontent.com/assets/6282045/13368621/478f39da-dcec-11e5-9457-4ef53f95b5aa.png)
When I add a new TCO, it gets grayed out, as if some values are not initialized. The TCO gets normal color when I resize it, select it, or do any of the actions in the context menu. I made the  `BBTCOView` behave the same way `PatternView` does (`m_needsUpdate` becomes true in the same spots), but it just won't behave the same way (maybe some signals and slots are missing).

Anyways, if someone here with a greater Qt Knowledge than me could help out, that'd be great.

tresf commented:
Is there a chance your new code checks `m_needsUpdate` before it's ever initialized?  It looks like the first draw may be getting missed because it's never set after first paint, but that's just a quick glance.

musikBear commented:
@Umcaruje Good job on the text :+1:

Fastigium commented:
@Umcaruje I think I found the problem. PatternView calls `setStyle( QApplication::style() )` in its constructor, while BBTCOView does not. Adding that line to the constructor of BBTCOView solves the problem for me.

It seems kind of hackish, though. Qt is supposed to set the style automatically, and it does, but delayed. On the second call to `BBTCOView::paintEvent()`, the style is set, but `m_needsUpdate` is already false. The `setStyle()` hack solves this, but I wonder where that delay comes from.

Umcaruje commented:
Thank you so much @Fastigium, this solves the problem. 

I have another question. To solve #2148, I need to connect a appropriate signal to a `update()` slot, so the patterns get redrawn.

Now, `Pattern.cpp` emits a `datachanged()` when its extended or shortened (https://github.com/LMMS/lmms/blob/master/src/tracks/Pattern.cpp#L477), so I should just connect that signal into the constructor `BBTCOView`, but I'm not sure exactly how to call Pattern from there.

```c++
connect( Pattern, SIGNAL( dataChanged() ),  this, SLOT( update() ) );
```
As expected, this code doesn't work.

Fastigium commented:
@Umcaruje You're welcome :smile:. As for the other question, I'm afraid using signals/slots isn't the right solution in this case. They connect between objects, not classes. That means that connections would have to be made between every Pattern object in the BBEditor and all BBTCOViews, or at least many of both.

A more promising road to fixing #2148 seems to lie in Pattern's `updateBBTrack()` method. I've just been trying to do so, but it's a bit more involved than I thought. If you don't mind, I'll see if I can fix this in a separate pull request.

Umcaruje commented:
>If you don't mind, I'll see if I can fix this in a separate pull request.

Ok, but this PR will make the matter worse, since it doesn't update the BBTCOView when moved, only when its resized. (All patterns are redrawn only when needed).

Umcaruje commented:
Ok, I think I'm done. SampleTCO's draw into pixmaps now too. I also refactored the drawing of the bar lines. In the end it all looks like this: 
![screenshot from 2016-02-28 01 12 19](https://cloud.githubusercontent.com/assets/6282045/13376281/591ffc80-ddb8-11e5-9cbd-5006df605a61.png)

@Wallacoloo @tresf @zonkmachine @michaelgregorius @Fastigium I would appreciate some code review/testing before I merge this.

tresf commented:
@Umcaruje 

An observation with the CSS, now that things are becoming consolidated... The common view colors can be defined in a single block to remove redundancies (normal CSS stuff).

If `TrackContentObjectView` accepted CSS that would simplify it even further.  I assume it is intentional, but they all share the same `qproperty-textColor` except `SampleTCOView`, so I just overrode that property in the `SampleTCOView` block.

```css

/ * common pattern colors */
PatternView, SampleTCOView, AutomationPatternView, BBTCOView {
	qproperty-mutedColor: rgb( 128, 128, 128 );
	qproperty-mutedBackgroundColor: rgb( 80, 80, 80 );
	qproperty-textColor: rgb( 255, 255, 255 );
	qproperty-textShadowColor: rgb( 0, 0, 0 );
	qproperty-selectedColor: rgb( 0, 125, 255 );
	qproperty-gradient: true;
}

/* automation pattern */
AutomationPatternView {
	background-color: rgb( 153, 175, 255 );
	color: rgb( 204, 215, 255 );
}

/* bb-pattern */
BBTCOView {
	background-color: rgb( 128, 182, 175 ); /* project may override */
	qproperty-selectedColor: rgb( 0, 125, 255 );
}

/* sample track pattern */
SampleTCOView {
	background-color: rgb( 74, 253, 133 );
	color: rgb( 187, 227, 236 );
	qproperty-textColor: rgb( 255, 60, 60 );
}

/* instrument pattern */
PatternView {
	background-color: rgb( 119, 199, 216 );
	color: rgb( 187, 227, 236 );
	font-size: 5pt;
}
```

tresf commented on include/BBTrack.h:
```diff
@@ -108,13 +111,21 @@ protected slots:
 
 protected:
 	void paintEvent( QPaintEvent * );
+	void resizeEvent( QResizeEvent * re )
```
Correct me if I'm wrong here (not great with C++), but if each class which inherits `TrackContentObjectView` simply calls `m_needsUpdate = true`' and then it's parent's `resizeEvent(...)`, then it the `m_needsUpdate` flag can probably be happen directly in `TrackContentObjectView::resizeEvent(...)`.  Also I don't think it's a `virtual void` in `SampleTCOView` if it contains code.

Umcaruje replied:
>Also I don't think the declaration of resizeEvent should be virtual void in SampleTCOView if it contains code.

https://github.com/LMMS/lmms/blob/de7d83d15897407b6300e2e3475ce6f7c62903a2/include/Pattern.h#L186

Its like that in Pattern.cpp as well. I'm not so good at c++, so I don't really get the difference between `virtual void` and `void`.

>I see m_needsUpdate is declared in each class. That might be redundant. See this: http://stackoverflow.com/a/14270779/3196753

I don't quite understand. I should define a private member variable `m_needsUpdate` in `TrackContentObjectView`, and then make functions that allow the child classes to access that variable?

Umcaruje commented:
>I assume it is intentional, but they all share the same qproperty-textColor except SampleTCOView, so I just overrode that one property in the SampleTCOView block.

Its not intentional, I just didn't do a final check :) Sample tracks currently don't get any text drawn on them, that should be fixed in another PR.

I'll make the CSS changes you proposed.

Fastigium commented:
@Umcaruje To fix #2148, try adding this to the constructor of `BBTCOView`:
```
connect( _tco->getTrack(), SIGNAL( dataChanged() ), this, SLOT( update() ) );
```
Signals/slots after all, but not between Pattern and BBTCOView. Turns out Pattern's `updateBBTrack()` method I mentioned makes the BBTrack that the Pattern belongs to emit a `dataChanged()` signal, and it's no problem to connect a BBTCOView to its track. I was a bit misled because there was no `emit` keyword, but today I learned signals always emit when called, with or without `emit` before them :blush:

Umcaruje commented:
@Fastigium it works! Thanks a lot.

tresf commented:
The diffs are outdated, so I'll reply in the PR comments directly... 

>  `virtual void` is like that in Pattern.cpp as well

Sorry, my own misunderstanding of it.  This article explains it: http://www.qtcentre.org/threads/864-diffrence-between-void-and-virtual-void, but it's fine as-is.

> I don't quite understand. I should define a private member variable `m_needsUpdate` in `TrackContentObjectView`, and then make functions that allow the child classes to access that variable?

Yes, exactly.  But, from a design perspective this would also assume you have a use for it inside TCOV by consolidating your common `paintEvent` code (assuming you encapsulate the variable, you could instead simply make it public) and then each variation to `paintEvent` can first call the [super-class first, then subsequent code](http://www.programiz.com/cpp-programming/function-overriding) and then add any additional logic afterward.  Of course, this conversation really is a duplicate of https://github.com/LMMS/lmms/pull/2574#issuecomment-186632362, so feel free to tell me to stop any any time. :smile:

Umcaruje commented:
Well, I implemented it so that `m_needsUpdate` is now a member variable of `TrackContentObjectView`. I implemented getter and setter functions `needsUpdate()` and `setNeedsUpdate()`. 

This is pretty much where I draw the line. I'd want to work on some new issues now and this code is in a much better state than before. As I said previously, complete consolidation of the code is currently out of the scope of this PR :)

tresf commented:
@Umcaruje well I don't think it harms to have an external API to force the redraw and agreed, out of scope.

I don't have a chance to compile and test this for a few days, so if someone else can give it the OK, that would be great.

tresf commented:
Things look good from here.  Good to merge.  These Piano roll "notches" in the BBEditor... are they a new feature?  I don't remember them before.


![image](https://cloud.githubusercontent.com/assets/6345473/13441862/73fdbbf8-dfc7-11e5-8c02-d1056a195245.png)

![image](https://cloud.githubusercontent.com/assets/6345473/13441875/8da268d8-dfc7-11e5-8f15-8bf093f063a8.png)


Also, unrelated to your PR, and I think already reported, but the clone pattern button doesn't impact the last BB track.

Umcaruje commented:
>These Piano roll "notches" in the BBEditor... are they a new feature? I don't remember them before.

They exist on the LMMS 1.1.3 I have installed on my PC:
![screenshot from 2016-03-01 23 49 46](https://cloud.githubusercontent.com/assets/6282045/13444762/6f5b2c8e-e008-11e5-95c7-629f23a2dff8.png)

Umcaruje commented:
Oh, don't merge yet: I just noticed the mute icons are misaligned from your pic: 
![screenshot from 2016-03-01 23 53 46](https://cloud.githubusercontent.com/assets/6282045/13444821/e12a8102-e008-11e5-9b31-495c7d4547f2.png)

I'll fix that tommorow, and then this will be ready to merge :)

Umcaruje commented:
![gifrecord_2016-03-02_215028](https://cloud.githubusercontent.com/assets/6282045/13475057/158a520c-e0c1-11e5-8a50-5ca15f786a12.gif)

The 'muted' pixmaps are now properly aligned, and also stay in place when you resize the track width. I also reduced their size to `14x14` so they don't clash with the text.

This is good to merge now :+1:

tresf commented:
:+1:

IvanMaldonado commented:
I know this is not the place to post this, but I didn't want to open a issue to inform a minor problem.

![captura de pantalla_2016-03-03_16-32-04](https://cloud.githubusercontent.com/assets/12549082/13512027/cf72ec6e-e15d-11e5-82b2-5499a06963e4.png)

I noticed there is a property in the style.css that should be moved.

Umcaruje commented:
>I noticed there is a property in the style.css that should be moved

Why?

IvanMaldonado commented:
I take every comment as a section, and the graphics that are changed by QToolButton are the same graphics changed by the properties under "all tool buttons" and it is currently under "smaller toolbars" besides QToolBar's properties.

![2](https://cloud.githubusercontent.com/assets/12549082/13512846/dd3cf074-e162-11e5-8fa6-b0a2c4276627.png)

Red: QToolButton
Blue: QToolBar

Umcaruje commented:
Then the comment should be moved, and not the property. Cuz, you see, we are overriding the colors on hover and also the border and other stuff. If that property was moved where you suggest, the look of the buttons wouldn't change.

Umcaruje commented:
No it shouldn't look like that, you just made the play and stop buttons square again...

Umcaruje commented:
Anyways, this conversation is entirely pointless on this PR and has 0 relevance to it. 

Issue a Pull Request or open an issue about this if you want, I'm locking this.
